proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1084 @@
+
+::
+
+  -- Given
+  class Foo a where
+    f :: F a
+    g :: G a
+
+  -- We have
+  f :: Foo a => F a
+  g :: Foo a => G a
+
+This is unchanged. But we add one exception: in type classes with exactly
+one method, the one method is linear in the class constraint
+
+::
+
+  -- Given
+  class Bar a where
+    h :: H a
+
+  -- We have
+  h :: Bar a %1 => H a
+
+Typing inference
+^^^^^^^^^^^^^^^^
+
+In a way, there's no need to worry about type inference: if a function
+``C %1 -> T``, with well-placed dictionaries, would be rejected, then
+``C %1 => T`` will be rejected as well. So understanding linear types
+is sufficient for the most part.
+
+But when there is an accepted assignment of type ``C %1 -> T``, it
+doesn't follow that the function of type ``C %1 => T`` will be
+accepted. Because GHC's typechecker doesn't make guesses.
+
+The one new rule introduced by this proposal is that when I want a
+linear constraint ``C`` and I've been given both a linear an an
+unrestricted ``C``, then this is considered ambiguous and raises a
+type error. See Section 6.3 of the paper_ for more details.
+
+To see why, consider this example
+
+::
+
+  class C
+  giveC :: (C => Int) -> Int
+  useC :: C %1 => Int
+
+  bad :: C %1 => (Int, Int)
+  bad = (giveC useC, useC)
+
+  bad' :: C %1 => (Int, Int)
+  bad' = (giveC useC, 0)
+
+In ``bad``, if the leftmost ``useC`` uses the linear ``C`` from the
+function signaturev, then ``bad`` would be rejected, it must used the
+unrestricted ``C`` from ``giveC``. But in ``bad'`` it must use the
+linear ``C`` instead. So this would force the leftmost ``useC`` to
+make a guess. Instead we reject both ``bad`` and ``bad'``.
+
+Equality constraints
+~~~~~~~~~~~~~~~~~~~~
+
+Given equality constraints are used for unification *only
+if they are unrestricted* (correspondingly, equality constraints
+generated by the type inference algorithm are unrestricted, as they
+have always been).
+
+The reason for this is that there is no clear semantics to make use of
+a linear equality constraint ``a ~ b`` as part of the unification
+algorithm (it's not that reasoning about linear equality is
+meaningless, but a unification or congruence conversion for linear
+equality isn't obvious to come up with, if someone has we're not
+aware). Nor does it feel like a true limitation as there is no example
+where a linear equality would be useful. It's really not worth the
+bother of trying to find a solution.
+
+Dupable classes
+^^^^^^^^^^^^^^^
+
+A new module ``GHC.Constraint.Linear`` is introduced (inlined
+alternative: bikeshed names, including the module name).
+
+This module exposes the following:
+
+::
+
+  data LinearityToken :: ZeroBitType
+
+  consumeLinearityToken :: LinearityToken %1 -> (# #)
+  dup2LinearityToken :: LinearityToken %1 -> (# LinearityToken, LinearityToken #)
+
+  data DupableClassModifier = Dupable
+
+Class declaration can be annotated with the ``%Dupable``
+
+::
+
+  class %Dupable <ctx> => <head> where
```

Comment:
We might be at odds here with the modifiers GHC proposal

Reply:
Did I place the modifier in the wrong place? I need to double-check.